### DISCUSSION TITLE
What's the point of the digital pins in the voltage ladder?

### DISCUSSION THREAD

**Discussion by FelixCodeStorage** [alf45tar/PedalinoMini#353] [Q&A]
Hi,
I don't understand why you need to use a digital pin in order to use the voltage ladder.
The 10k Ohms pull up resistor from 3,3V to the analog pin and the resistors of the voltage ladder form a voltage divider which should easily be measurable with only the analog pin.

Do you use the digital Pins as 3,3 V sources ?

The voltage divider between 10k Ohms and 330 Ohms should result in 3,2V analog Pin voltage while a divider between the digital Pins 2k Ohms and 330 Ohms should result in 2,8V which makes more sense.

When my assumption is correct you could easily free up 6 digital pins and increase the accuracy by using a 2k Ohm analog Pin pull up resistor. 

I attached a drawing of the circuit I had in mind.

Thanks for your answer
![photo_2022-06-27_23-37-52](https://user-images.githubusercontent.com/66059446/176039899-ab0d6887-daf3-481e-b1d0-c06a43d4cb66.jpg)
.

**Comment by alf45tar** [comment]
I didn't tested the circuit but it should work. You do not need to change the firmware just disconnect the digital pin.
A firmware change is requested to reuse the digital pin but this is another story.
I will test you circuit and if it works I will create 2 new modes called Ladder+Momentary and Ladder+Latch.

**Reply by FelixCodeStorage** [reply]
I think it would be better if you would add 6 more pedals in the pedals tab because then you could have 36 Pedals + 3 Jog wheels, as the jog wheels only need 3 digital Pins in order to work.
Having two jog wheels would be very awesome to control virtual guitar effects with parameters. 

This way it would also be easier to understand.

**Reply by alf45tar** [reply]
I tested the circuit and it works pretty well. I will not change the 10k pull up resistors but you are free to customise them as you prefer.
I will not add 6 more pedals because the logic behind is different. PedalinoMini has 6 equal ports and each of them can be configured in different way. Each port is a pedal and each pedal can have multiple "buttons".

**Reply by FelixCodeStorage** [reply]
The ESP32 ADC is very inaccurate close to the 3,3V, so this could become a problem...

Mhh... I understand.
Can you think of another way to use the now free 3 digital Pins with Jog wheels? Maybe an option to configure the analog and the digital pins of a Pedal in a different way?

**Reply by alf45tar** [reply]
Use different (greater) resistors value to equally divide the resolutions. I suggest the following reading.

http://www.ignorantofthings.com/2018/07/the-perfect-multi-button-input-resistor.html

**Reply by alf45tar** [reply]
Can you connect a jog wheel to another port? Do you want to build a PedalinoMini with 36 switches and 1 jog wheel?

**Reply by FelixCodeStorage** [reply]
> Use different (greater) resistors value to equally divide the resolutions. I suggest the following reading.
> 
> http://www.ignorantofthings.com/2018/07/the-perfect-multi-button-input-resistor.html

yes this will work, but what I wanted to say is that the internal ref voltage of the esp32 is 1,1V.
So in order to read up to 3,3V the input gets amplified which can cause distortions.
When you equally space the resistors is should not be problem.

**Reply by FelixCodeStorage** [reply]
> Can you connect a jog wheel to another port? Do you want to build a PedalinoMini with 36 switches and 1 jog wheel?

I don't want to build it, but I think some people would like to have the possibility to do it.
When you use 2 expression pedals, 2 analog pins are already gone. Having these + 18 Buttons + 2 jog wheels would be very nice.

With these possibilities, the pedallino could also be a on desk device for audio mixing when you connect volume sliders to the analog pins.

I want to push you a little bit to think big, because everything you did so far ist simply amazing.

**Reply by alf45tar** [reply]
With the additional digital pin on the same port you can attach a wah wah pedal with the switch like the cry baby.

**Comment by alf45tar** [comment]
Just released 2.4.9 with a new pedal mode: Analog+Switch.
Schematic is here https://github.com/alf45tar/PedalinoMini/blob/master/images/Schematic_PedalinoMini_Sheet-3.svg

**Reply by FelixCodeStorage** [reply]
I just installed it, is it on purpose that I can only select one button with this config?
If not, it would make sense if one button would be the analog one and one the digital one.
![image](https://user-images.githubusercontent.com/66059446/178112830-1588a7e6-5ef2-4dff-9998-abcaf6ebaf11.png)

**Reply by alf45tar** [reply]
Yes, it is. Right now you can have only one default led for both.
It can be overwritten on actions of course.
